pkg/pillar/agentbase/agent.go

```diff
@@ -15,54 +15,114 @@ import (
+	}
+}
+
 // Context - a struct that represents a general agent context
 type Context struct {
-	PubSub       *pubsub.PubSub
-	Logger       *logrus.Logger
-	Log          *base.LogObject
-	CLIParams    CliParams
-	ErrorTime    time.Duration
-	WarningTime  time.Duration
-	AgentName    string
-	NeedWatchdog bool
+	pubSub       *pubsub.PubSub
+	logger       *logrus.Logger
+	log          *base.LogObject
+	cliParams    CliParams
+	errorTime    time.Duration
+	warningTime  time.Duration
+	agentName    string
+	needWatchdog bool
+	needPidFile  bool
+	arguments    []string
+}
+
+// CLIParams returns CliParams
+func (ctx *Context) CLIParams() CliParams {
+	return ctx.cliParams
+}
+
+// Logger returns Logger
+func (ctx *Context) Logger() *logrus.Logger {
+	return ctx.logger
+}
+
+// InitAgentBaseContext initialize Context with defined options
+func InitAgentBaseContext(logger *logrus.Logger, log *base.LogObject, agentName string, opts ...ContextOpt) *Context {
+	ctx := &Context{
+		logger:    logger,
+		log:       log,
+		agentName: agentName,
+	}
+	for _, opt := range opts {
+		opt(ctx)
+	}
+	return ctx
 }
 
 // processCLIFlags - Add flags common to all agents
-func processCLIFlags(agentBase AgentBase) {
-	debugPtr := flag.Bool("d", false, "Debug flag")
-	agentBase.AddAgentSpecificCLIFlags()
-	flag.Parse()
+func processCLIFlags(agentBase AgentBase) error {
 	ctx := agentBase.AgentBaseContext()
-	ctx.CLIParams.DebugOverride = *debugPtr
-	agentBase.ProcessAgentSpecificCLIFlags()
+	flagSet := flag.NewFlagSet(ctx.agentName, flag.ExitOnError)
+	debugPtr := flagSet.Bool("d", false, "Debug flag")
+	agentBase.AddAgentSpecificCLIFlags(flagSet)
+	if err := flagSet.Parse(ctx.arguments); err != nil {
+		return err
+	}
+	ctx.cliParams.DebugOverride = *debugPtr
+	agentBase.ProcessAgentSpecificCLIFlags(flagSet)
+	return nil
 }
 
-// Run - a general run function that will handle all of the common code for agents
-func Run(agentSpecificContext AgentBase) {
+// Init - a general init function that will handle the common code for agents
+func Init(agentSpecificContext AgentBase) {
 	ctx := agentSpecificContext.AgentBaseContext()
-	logger := ctx.Logger
-	log := ctx.Log
-	processCLIFlags(agentSpecificContext)
-	debugOverride := ctx.CLIParams.DebugOverride
+	logger := ctx.logger
+	log := ctx.log
+	if err := processCLIFlags(agentSpecificContext); err != nil {
+		log.Fatal(err)
+	}
+	debugOverride := ctx.cliParams.DebugOverride
 	if debugOverride {
 		logger.SetLevel(logrus.TraceLevel)
 	} else {
 		logger.SetLevel(logrus.InfoLevel)
 	}
-	if err := pidfile.CheckAndCreatePidfile(log, ctx.AgentName); err != nil {
-		log.Fatal(err)
+	if ctx.needPidFile {
+		if err := pidfile.CheckAndCreatePidfile(log, ctx.agentName); err != nil {
+			log.Fatal(err)
+		}
 	}
-	log.Functionf("Starting %s\n", ctx.AgentName)
-	if ctx.NeedWatchdog {
-		ctx.PubSub.StillRunning(ctx.AgentName, ctx.WarningTime, ctx.ErrorTime)
+	log.Functionf("Starting %s\n", ctx.agentName)
+	if ctx.needWatchdog {
```

Comment:
out of the curiosity: what is the zededa coding style? structure members are camel style which do start with a small letter?

Reply:
Coding style is mostly decided by Go (https://go.dev/blog/gofmt) and partially by Yetus (like enum names etc). 
And the capitalization is related to field exporting, see: https://golangbyexample.com/exported-unexported-fields-struct-go/

Reply:
But not everything is enforced by the tools that we use inside the workflow, some of that is a convention in Go (like preference to use camel case). This a good article on the topic: https://betterprogramming.pub/naming-conventions-in-go-short-but-descriptive-1fa7c6d2f32a

Reply:
Thanks Milan, good to know.